### PR TITLE
Drop `S3ExpressSigner` and override session token name via `RuntimePlugin`

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
In response to https://github.com/smithy-lang/smithy-rs/pull/3455#issuecomment-1973771991, it's clear that we need to iterate more on the potential signing API.

This PR takes a different approach where `S3ExpressSigner` is dropped and a session token name override, if any, is placed into and retrieved from `ConfigBag`, which is used within `SigV4Signer::sign_http_request`.

## Testing
Existing tests in CI

A semver failure in CI can be ignored; A public API SigV4Signer::sign_http_request has been removed that only existed in the branch.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/rust-runtime/aws-runtime/src/auth.rs [hunk 1]
```diff
@@ -199,7 +209,8 @@ fn extract_field_from_endpoint_config<'a>(
         .and_then(|config| config.get(field_name))
 }
 
-fn apply_signing_instructions(
+/// Applies the instructions to the given `request`.
+pub fn apply_signing_instructions(
```
Comment (ysaito1001) [inline aws/rust-runtime/aws-runtime/src/auth.rs:213]: I attempted to use `signing_instructions.apply_to_request_http0x(request)` in `S3ExpressSigner::sign_http_request`, but the variable `request` was passed by a caller whose type is `&mut HttpRequest`.
Calling  `.try_into_http02x()` on `request` within `S3ExpressSigner::sign_http_request` would convert that to the underlying type, but `request` needed to continue being used, expecting it to be orchestrator's `HttpRequest`.

That's the rationale for making this function `pub`.

File: aws/rust-runtime/aws-runtime/src/auth.rs [hunk 2]
```diff
@@ -75,6 +75,52 @@ impl Default for SigningOptions {
...
 
+pub(crate) type SessionTokenNameOverrideFn = Box<
+    dyn Fn(&SigningSettings, &ConfigBag) -> Result<Option<&'static str>, BoxError>
+        + Send
+        + Sync
+        + 'static,
+>;
+
+/// Custom config that provides the alternative session token name for [`SigningSettings`]
+pub struct SessionTokenNameOverride {
```
Comment (jdisanti) [inline aws/rust-runtime/aws-runtime/src/auth.rs:86]: Nit: should prefix this with SigV4

**Review by jdisanti — APPROVED** [review]
Looks great!

File: aws/rust-runtime/aws-runtime/src/auth.rs (hunk 2 above)
Comment (ysaito1001) [reply aws/rust-runtime/aws-runtime/src/auth.rs]: Renamed in 0193d49

**Review by jdisanti — APPROVED** [review]

**Pull request merged**